src/oracles/CIXPythAdapter.sol

```diff
@@ -0,0 +1,219 @@
+    // - Reduce the parameter with geometric average calculation.
+    //   Calculate geometric average price according to the formula
+    //   price = c * (price1 ^ +-weight1) * (price2 ^ +-weight2) * ... * (priceN ^ +-weightN)
+    // - Keep track of minimum publish time
+
+    // Declare _accum as c
+    int128 _accum = _convertE8To64x64(_config.cE8);
+
+    for (uint256 i = 0; i < _len; ) {
+      // Get price from Pyth
+      PythStructs.Price memory _priceStruct = pyth.getPriceUnsafe(_config.pythPriceIds[i]);
+      uint256 _priceE8 = _convertToUint256(_priceStruct, 8);
+
+      // Accumulate the _accum with (priceN ^ +-weightN)
+      _accum = _accumulateWeightedPrice(_accum, _priceE8, _config.weightsE8[i], _config.usdQuoteds[i]);
+
+      // Update publish time, with minimum _price.publishTime
+      if (i == 0) {
+        _publishTime = _priceStruct.publishTime;
+      } else {
+        _publishTime = HMXLib.min(_publishTime, _priceStruct.publishTime);
+      }
+
+      unchecked {
+        ++i;
+      }
+    }
+
+    // 3. Convert the final result to uint256 in e30 basis
+    _priceE30 = _convert64x64ToE8(_accum) * 1e22;
+
+    return (_priceE30, _publishTime);
+  }
+
+  function getConfigByAssetId(bytes32 _assetId) external view returns (ICIXPythAdapter.CIXPythPriceConfig memory) {
+    return configs[_assetId];
+  }
+
+  /**
+   * Setter
+   */
+  /// @notice Set new Pyth contract address.
+  /// @param _newPyth New Pyth contract address.
+  function setPyth(address _newPyth) external onlyOwner {
+    emit LogSetPyth(address(pyth), _newPyth);
+    pyth = IReadablePyth(_newPyth);
+  }
+
+  /// @notice Set the Pyth price id for the given asset.
+  /// @param _assetId The HMX asset id to set. (not Pyth priceId)
+  /// @param _cE8 A magic constant. Need to be recalculate every time the weight is changed.
+  /// @param _pythPriceIds An array of pythPriceIds. This should be relative to _weightsE8.
+  /// @param _weightsE8 An array of weights of certain asset in e8 basis. This should be relative to _pythPriceIds.
+  function setConfig(
+    bytes32 _assetId,
+    uint256 _cE8,
+    bytes32[] memory _pythPriceIds,
+    uint256[] memory _weightsE8,
+    bool[] memory _usdQuoteds
+  ) external onlyOwner {
+    ICIXPythAdapter.CIXPythPriceConfig memory _config;
+
+    // 1. Validate params
+
+    uint256 _len = _pythPriceIds.length;
+    // Validate length
+    {
+      if (_len != _weightsE8.length || _len != _usdQuoteds.length) revert CIXPythAdapter_BadParams();
+      if (_cE8 == 0) revert CIXPythAdapter_BadParams();
+    }
+
+    // Validate weight and price id
+    {
+      uint256 _weightSum;
+      for (uint256 i = 0; i < _len; ) {
+        // Accum weight sum
+        _weightSum += _weightsE8[i];
+
+        // Sanity check for price id
+        pyth.getPriceUnsafe(_pythPriceIds[i]);
+
+        unchecked {
+          ++i;
+        }
+      }
+
+      if (_weightSum != 1e8) revert CIXPythAdapter_BadWeightSum();
+    }
+
```

Comment:
Should have a variation check as C is a critical variable which can change the index prices.